File: modules/webrtc-core/src/main/java/com/shterneregen/securelan/webrtc/runtime/RtcEngineProvider.java
package com.shterneregen.securelan.webrtc.runtime;

public final class RtcEngineProvider {
    private RtcEngineProvider() {
    }

    public static RtcEngine createDefault() {
        try {
            return new WebRtcJavaEngine();
        } catch (Throwable error) {
            error.printStackTrace();
            return new NoOpRtcEngine(buildFailureMessage(error));
        }
    }

    private static String buildFailureMessage(Throwable error) {
        Throwable root = error;
        while (root.getCause() != null && root.getCause() != root) {
            root = root.getCause();
        }

        String outer = error.getClass().getSimpleName()
                + (error.getMessage() == null ? "" : ": " + error.getMessage());
        String inner = root.getClass().getSimpleName()
                + (root.getMessage() == null ? "" : ": " + root.getMessage());

        return "Failed to initialize webrtc-java provider: "
                + outer
                + " | root cause: "
                + inner;
    }
}

File: modules/webrtc-core/src/main/kotlin/com/shterneregen/securelan/webrtc/runtime/RtcEngineProvider.kt
package com.shterneregen.securelan.webrtc.runtime

object RtcEngineProvider {
    @JvmStatic
    fun createDefault(): RtcEngine = try {
        WebRtcJavaEngine()
    } catch (error: Throwable) {
        error.printStackTrace()
        RtcFileLogger.error("Failed to initialize webrtc-java provider", error)
        NoOpRtcEngine(buildFailureMessage(error))
    }

    @JvmStatic
    fun createDefault(preferredAudioCaptureDeviceId: String?, preferredVideoCaptureDeviceId: String?): RtcEngine = try {
        WebRtcJavaEngine(preferredAudioCaptureDeviceId, preferredVideoCaptureDeviceId)
    } catch (error: Throwable) {
        error.printStackTrace()
        RtcFileLogger.error("Failed to initialize webrtc-java provider", error)
        NoOpRtcEngine(buildFailureMessage(error))
    }

    private fun buildFailureMessage(error: Throwable): String {
        var root = error
        while (root.cause != null && root.cause !== root) {
            root = root.cause!!
        }

        val outer = error.javaClass.simpleName + (error.message?.let { ": $it" } ?: "")
        val inner = root.javaClass.simpleName + (root.message?.let { ": $it" } ?: "")

        return "Failed to initialize webrtc-java provider: $outer | root cause: $inner"
    }
}
